library/core/src/time.rs

```diff
@@ -794,44 +863,85 @@ impl Duration {
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f32(1e-20);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// let res = Duration::try_from_secs_f32(4.2e-7);
+    /// assert_eq!(res, Ok(Duration::new(0, 419)));
+    /// let res = Duration::try_from_secs_f32(2.7);
+    /// assert_eq!(res, Ok(Duration::new(2, 700_000_047)));
+    /// let res = Duration::try_from_secs_f32(3e10);
+    /// assert_eq!(res, Ok(Duration::new(30_000_001_024, 0)));
+    /// // subnormal float:
+    /// let res = Duration::try_from_secs_f32(f32::from_bits(1));
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    /// // conversion uses truncation, not rounding
+    /// let res = Duration::try_from_secs_f32(0.999e-9);
+    /// assert_eq!(res, Ok(Duration::new(0, 0)));
+    ///
+    /// let res = Duration::try_from_secs_f32(-5.0);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f32(f32::NAN);
+    /// assert!(res.is_err());
+    /// let res = Duration::try_from_secs_f32(2e19);
+    /// assert!(res.is_err());
     /// ```
     #[unstable(feature = "duration_checked_float", issue = "83400")]
     #[inline]
-    pub const fn try_from_secs_f32(secs: f32) -> Result<Duration, FromSecsError> {
-        const MAX_NANOS_F32: f32 = ((u64::MAX as u128 + 1) * (NANOS_PER_SEC as u128)) as f32;
-        let nanos = secs * (NANOS_PER_SEC as f32);
-        if !nanos.is_finite() {
-            Err(FromSecsError { kind: FromSecsErrorKind::NonFinite })
-        } else if nanos >= MAX_NANOS_F32 {
-            Err(FromSecsError { kind: FromSecsErrorKind::Overflow })
-        } else if nanos < 0.0 {
-            Err(FromSecsError { kind: FromSecsErrorKind::Negative })
-        } else {
-            let nanos = nanos as u128;
-            Ok(Duration {
-                secs: (nanos / (NANOS_PER_SEC as u128)) as u64,
-                nanos: (nanos % (NANOS_PER_SEC as u128)) as u32,
-            })
+    pub const fn try_from_secs_f32(secs: f32) -> Result<Duration, FromFloatSecsError> {
+        const MANT_BITS: i16 = 23;
+        const MANT_MASK: u32 = (1 << MANT_BITS) - 1;
+        const EXP_BITS: i16 = 8;
+        const EXP_MASK: u32 = (1 << EXP_BITS) - 1;
+        const OFFSET: i16 = 41;
+        const DIV1: u64 = 1u64 << MANT_BITS;
+        const DIV2: u128 = 1u128 << (MANT_BITS + OFFSET);
+
+        if secs.is_sign_negative() {
+            return Err(FromFloatSecsError { kind: FromFloatSecsErrorKind::Negative });
         }
+
+        let bits = secs.to_bits();
+        let mant = (bits & MANT_MASK) | (MANT_MASK + 1);
+        let exp = ((bits >> MANT_BITS) & EXP_MASK) as i16 - 127;
+
+        // note: max `f32` for `exp` equal to -31 is 9.31e-10, which is less than 1 ns
+        let (secs, nanos) = match exp {
+            -127..=-31 => (0, 0),
+            -30..=-1 => {
+                let t = u64::from(mant) << (OFFSET + exp);
+                let nanos = (u128::from(NANOS_PER_SEC) * u128::from(t)) / DIV2;
```

Comment:
I think its probably better to replace this division with a shift. Right now the relationship between `<< OFFSET` above and the offset on this line is not all that trivial to decipher.

(Alternatively replace the `<<` above with a multiplication)